elks/arch/i86/drivers/block/ata-cf.c

```
@@ -0,0 +1,206 @@
+/**********************************************************************
+ * ELKS ATA-CF driver
+ *
+ * Please note that this driver does not currently support partitions.
+ *
+ * This driver is largely based on Greg Haerr's SSD driver.
+ *
+ * Ferry Hendrikx, June 2025
+ **********************************************************************/
+
+#include <linuxmt/config.h>
+#include <linuxmt/kernel.h>
+#include <linuxmt/errno.h>
+#include <linuxmt/debug.h>
+#include <arch/ata.h>
+#include "ata.h"
+#include "ata-cf.h"
+#include "blk.h"
+
+
+/**********************************************************************
+ * device initialisation
+ **********************************************************************/
+
+jiff_t ata_cf_timeout;
+
+char ata_cf_initialized;
+sector_t ata_cf_num_sects[NUM_DRIVES];     /* max # sectors on SSD device */
+static int access_count[NUM_DRIVES];
+
+
+static int ata_cf_ioctl(struct inode *inode, struct file *file, unsigned int cmd, unsigned int arg);
+static int ata_cf_open(struct inode *, struct file *);
+static void ata_cf_release(struct inode *, struct file *);
+
+static struct file_operations ata_cf_fops = {
+    NULL,                       /* lseek */
+    block_read,                 /* read */
+    block_write,                /* write */
+    NULL,                       /* readdir */
+    NULL,                       /* select */
+    ata_cf_ioctl,               /* ioctl */
+    ata_cf_open,                /* open */
+    ata_cf_release              /* release */
+};
+
+
+/**********************************************************************
+ * ATA-CF functions
+ **********************************************************************/
+
+void INITPROC ata_cf_init(void)
+{
+    sector_t sectors;
+    int i;
+
+    // register device
+
+    if (register_blkdev(MAJOR_NR, DEVICE_NAME, &ata_cf_fops) == 0)
+        blk_dev[MAJOR_NR].request_fn = DEVICE_REQUEST;
```

If register_blkdev fails, it might be better to just return, rather than to continue into driver initialization with ata_reset etc. However, the only reason register_blkdev can fail is if the MAJOR_NR is already in use, which it would seem could only happen if BIOSHD was also configured, so this isn't really much of an issue.